src/xdp-utils.c

```diff
@@ -572,6 +573,62 @@ cleanup_temp_file (void *p)
   g_free (*pp);
 }
 
+static GKeyFile *
+spawn_validator (const char* const* args,
+                 int                fd)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GKeyFile) key_file = NULL;
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) process = NULL;
+  int stdin_fd = -1;
+  int stdout_fd = -1;
+  g_autofree char *stdout_path = NULL;
+
+  stdout_fd = memfd_create("spawn-validator-stdout", 0);
+
+  if (stdout_fd == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_warning ("Failed to spawn subprocess: memfd_create: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  stdin_fd = dup (fd);
+  /* Ensure that we read from the beginning of the file */
+  lseek(stdin_fd, 0, SEEK_SET);
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_take_stdout_fd (launcher, stdout_fd);
+  g_subprocess_launcher_take_stdin_fd (launcher, stdin_fd);
```

Comment:
> The passed fd is noted but will not be touched in the current process. It is therefore necessary that it be kept open by the caller until the subprocess is spawned. The file descriptor will also not be explicitly closed on the child side, so it must be marked O_CLOEXEC if that’s what you want.

As I understand this, it doesn't take ownership of the fd. It only needs to be open until `g_subprocess_launcher_spawnv`. That means you shouldn't dup stdin and you have to close stdout_fd in this function.

Reply:
I had a really hard time to understand how `take_stdout_fd/take_stdin_fd()` worked, but the docs are not clear/correct about it. I was mostly confused about `take` in the name where the docs say that the fd is not taken.

Anyhow, the fds are closed once the launcher is dropped.  Looking at the docs for `g_subprocess_launcher_close()` and code https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gsubprocesslauncher.c#L647 we can see that.

Reply:
In the child, not in the parent.